File: apps/api/src/kimy/schemas/reports.py
from __future__ import annotations

from pydantic import BaseModel


class SubmissionReportRow(BaseModel):
    submission_id: str
    program_code: str
    program_name: str
    title: str
    chapter: str | None
    status: str
    student_name: str
    advisor_name: str | None
    decimal_grade: float | None
    total_percentage: float | None
    advisor_fit_score: float | None
    advisor_fit_alert: bool
    latest_version: int
    created_at: str


class SubmissionsReport(BaseModel):
    program_id: str | None
    status: str | None
    rows: list[SubmissionReportRow]
    total: int


class ProgramRollupRow(BaseModel):
    program_id: str
    program_code: str
    program_name: str
    submissions_count: int
    average_grade: float | None
    plagiarism_alerts: int
    fit_alerts: int


class ProgramsRollup(BaseModel):
    rows: list[ProgramRollupRow]

File: apps/web/src/lib/api/reports.ts
"use server";

import { apiFetch } from "@/lib/api/client";

export type SubmissionReportRow = {
  submission_id: string;
  program_code: string;
  program_name: string;
  title: string;
  chapter: string | null;
  status: string;
  student_name: string;
  advisor_name: string | null;
  decimal_grade: number | null;
  total_percentage: number | null;
  advisor_fit_score: number | null;
  advisor_fit_alert: boolean;
  latest_version: number;
  created_at: string;
};

export type SubmissionsReport = {
  program_id: string | null;
  status: string | null;
  rows: SubmissionReportRow[];
  total: number;
};

export type ProgramRollupRow = {
  program_id: string;
  program_code: string;
  program_name: string;
  submissions_count: number;
  average_grade: number | null;
  plagiarism_alerts: number;
  fit_alerts: number;
};

export type ProgramsRollup = {
  rows: ProgramRollupRow[];
};

export async function fetchSubmissionsReport(
  filters: { program_id?: string; status?: string } = {},
): Promise<SubmissionsReport> {
  const qs = new URLSearchParams();
  if (filters.program_id) qs.set("program_id", filters.program_id);
  if (filters.status) qs.set("status", filters.status);
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return apiFetch<SubmissionsReport>(`/api/v1/reports/submissions${suffix}`);
}

export async function fetchProgramsRollup(): Promise<ProgramsRollup> {
  return apiFetch<ProgramsRollup>("/api/v1/reports/programs");
}
